### PR TITLE
feat(flux): add support for renovating OCIRespository

### LINKED ISSUE #19034: Adding support for `OCIRepository` for `flux` manager
### What would you like Renovate to be able to do?

I would like to add support for updating `OCIRepository` resources for the `flux` manager.
This will be helpful for me since a lot of my resources in the flux cluster are from oci repos.

### If you have any ideas on how this should be implemented, please tell us here.

This follows closely with https://github.com/renovatebot/renovate/issues/18781 but a few minor differences in the `OCIRepository` spec and instead of using Git Datasource it uses Docker.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by onedr0p**
Signed-off-by: Devin Buhl <devin@buhl.casa>

## Changes

Supports renovating https://fluxcd.io/flux/components/source/ocirepositories/ with either `tag` or `digest`

This PR does not address `HelmRepository` OCI only `OCIRepository`

## Context

- closes https://github.com/renovatebot/renovate/issues/19034
- old PR #18999

This adds support for renovating Flux's `OCIRepository`

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/flux/index.ts
Comment (viceice) [inline lib/modules/manager/flux/index.ts:12]: please revert this, you probably need some extensions for your vscode, so it will follow our code style

checkout our recommend extensions  inside `.vscode` folder

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
please reuse old PR in future. so we don't loose discussion history. you can force push in cases when a merge commit is to complicated to solve conflicts.

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:34]: What happens when the `OCIRepository` fields have a `tag` **and** a `digest` key?

File: lib/modules/manager/flux/extract.spec.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.spec.ts:492]: The Flux docs say you can have both a tag and digest but digest will take precedence.

File: lib/modules/manager/flux/readme.md
Comment (onedr0p) [reply lib/modules/manager/flux/readme.md:34]: According to the docs it seems digest will take precedent.

https://fluxcd.io/flux/components/source/ocirepositories/#digest-example

File: lib/modules/manager/flux/index.ts
Comment (onedr0p) [reply lib/modules/manager/flux/index.ts:12]: I'm using the vscode devcontainer and also ran `yarn run lint-fix` before I commit this. Shouldn't both cases here cause that not to happen?

**Comment by onedr0p** [comment]
@viceice this won't close #18509 but I do plan to add support for that in another PR. I'll open an issue for OCIRepository and update the PR description to link to it.

File: lib/modules/manager/flux/__fixtures__/ociSource.yaml
Comment (onedr0p) [reply lib/modules/manager/flux/__fixtures__/ociSource.yaml:1]: I am not sure how to do this, since the file is needed in the `extractAllPackageFiles()` test in `extract.spec.ts`

File: lib/modules/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/flux/readme.md:34]:
```suggestion
The `flux` manager only updates `OCIRepository` fields that have a `digest` or `tag` key.
If `tag` _and_ `digest` exist then `digest` takes precedence.
```

We follow the "one sentence per line pattern" in our docs. Plus we should be explicit about the fields, so people don't have to guess what "both exist" means. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Small docs change.

File: lib/modules/manager/flux/__fixtures__/ociSource.yaml
Comment (onedr0p) [reply lib/modules/manager/flux/__fixtures__/ociSource.yaml:1]: According to https://github.com/renovatebot/renovate/blob/03d818a1eacbc6dbb4cf6a19b19360fee021cae9/docs/development/adding-a-package-manager.md

> The extractAllPackageFiles function takes an array of filenames as input. It returns an array of filenames and dependencies.

This would not be possible with an inline fixture, but please point me in the right place.

Comment (viceice) [reply lib/modules/manager/flux/__fixtures__/ociSource.yaml:1]: yes. fs should be mocked in tests like here:
https://github.com/renovatebot/renovate/blob/11a96b72a6b2564576edbc50086dd3cb28b4a390/lib/modules/manager/bundler/extract.spec.ts#L3
https://github.com/renovatebot/renovate/blob/11a96b72a6b2564576edbc50086dd3cb28b4a390/lib/modules/manager/bundler/extract.spec.ts#L7
https://github.com/renovatebot/renovate/blob/11a96b72a6b2564576edbc50086dd3cb28b4a390/lib/modules/manager/bundler/extract.spec.ts#L37

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:221]: we should support both too. so renovate can update digest, when the tag points to a new digest, as the docker tags are mutable.

that would be same behavior as we do on other docker based managers

Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:221]: I would normally agree but with Flux you can only have one or the other and if both tag and digest are present tag is ignored and digest is used.

Comment (viceice) [reply lib/modules/manager/flux/extract.ts:221]: thats same behavior as docker, digest is preferred.

as aa user i would expect that renovate will follow the tags to update the digest, when both are defined. 

currently renovate would update to digest of latest tag. that would be unexpected by most users.

@rarkins @JamieMagee WDYT?

Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:221]: It does appear we can have a digest in the `.spec.ref.tag` field, e.g.

```
v1.8.2@sha256:761c3189c482d0f1f0ad3735ca05c4c398cae201d2169f6645280c7b7b2ce6fc
```

but `.spec.ref.digest` can only have the digest, e.g.

```
sha256:761c3189c482d0f1f0ad3735ca05c4c398cae201d2169f6645280c7b7b2ce6fc
```

Does this help clear things up? If so I can add some tests to verify the former works.

File: lib/modules/manager/flux/__fixtures__/ociSource.yaml
Comment (onedr0p) [reply lib/modules/manager/flux/__fixtures__/ociSource.yaml:1]: That makes sense, but I am not sure I can retrofit this change into this PR as it is growing already, would it be better to submit a refactoring PR to address this fixture and the others after this PR is merged?

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts]: Added support for renovating tag and digest in `.spec.tag.ref` with a value `v1.8.2@sha256:761c3189c482d0f1f0ad3735ca05c4c398cae201d2169f6645280c7b7b2ce6fc`

File: lib/modules/manager/flux/__fixtures__/ociSource.yaml
Comment (onedr0p) [reply lib/modules/manager/flux/__fixtures__/ociSource.yaml:1]: Resolving this, please feel free to comment if what I mentioned is not ideal.

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts]: Resolving this as it needs a re-review.

**Review by HonkingGoose — DISMISSED** [review]

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:224]: if you don't like to support tag and digest as mentioned before you should log a warning, so users get a message that it's not supported. otherwise we'll get wrong bug reports about unexpected behavior.

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:24]: i would prefer to use the getDep function of dockerfile manager

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:215]: If only `digest` is set I have updated the currentValue to force being `latest` as that is what Flux assumes when only using `digest`.

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts]: I have made those changes, I now have it using `getDep()`.

**Comment by onedr0p** [comment]
@viceice I believe I am all good to go here unless you had suggestions on the most recent changes.

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:217]:
```suggestion
```
docker datasource will use latest if there's no current value

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:216]: i would still log a debug message when tag is also found.
should be something like `digest and tag found, ignoring tag`

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:220]:
```suggestion
          const containerTag = getDep(container + ':' + resource.spec.ref.tag);
```
use template string instead of concatenation

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:222]:
```suggestion
            '{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';
```
i don't think we want to have the depName inside the tag

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:226]:
```suggestion
          dep.replaceString = resource.spec.ref.tag;
```

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:220]:
```suggestion
          dep = getDep(container + ':' + resource.spec.ref?.tag);
```
why not using the returned value directly. also pass the no auto-replace template flag.

i think it's also a good idea to pass registry aliases.

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:216]: Fixed in upcoming commit

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:219]: Fixed in upcoming commit

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:221]: Fixed in upcoming commit

File: lib/modules/manager/flux/extract.ts
Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:219]: @viceice 

1. `dep` is a `const` that is used in all the conditional blocks, what you suggest won't work. What do you mean by that?
2. I tried pass the registry aliases but I received the error:

```
  ● modules/manager/flux/extract › extractPackageFile() › extracts OCIRepository with a tag

    TypeError: Cannot read properties of undefined (reading 'registryAliases')

      221 |           }
      222 |         } else if (resource.spec.ref?.tag) {
    > 223 |           const containerTag = getDep(`${container}:${resource.spec.ref.tag}`, false, config.registryAliases);
          |                                                                                              ^
      224 |           dep.autoReplaceStringTemplate =
      225 |             '{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';
      226 |           dep.currentDigest = containerTag.currentDigest;

      at resolveResourceManifest (lib/modules/manager/flux/extract.ts:223:94)
      at extractPackageFile (lib/modules/manager/flux/extract.ts:264:14)
      at Object.<anonymous> (lib/modules/manager/flux/extract.spec.ts:519:40)
```

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [reply lib/modules/manager/flux/extract.ts]: please reduce noise and read our contributing guide[^1]

[^1]: https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [reply lib/modules/manager/flux/extract.ts]: ok, remove `registryAliases` for now, can be done later and would need a lot more changes. you can checkout the dockerfile manager, or just search for `registryAliases` in codebase

Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts]: I did search around a bit on that while I was trying to make it work but ran into those issues. Maybe a small refactor is needed for it to fall into place.

**Comment by onedr0p** [comment]
@viceice anyhing else you want me to cover in this PR?

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:222]:
```suggestion
          dep = getDep(
```
make dep a let instead of const, use same for digest above.

the get dep will do almost all the work below, just disable the auto-replace.

we'll add the registry aliases feature in a later PR.

Comment (onedr0p) [reply lib/modules/manager/flux/extract.ts:222]:
```
Type 'PackageDependency<Record<string, any>>' is not assignable to type 'PackageDependency<FluxManagerData>'.
Property 'components' is missing in type 'Record<string, any>' but required in type 'FluxManagerData'.
```

Any suggestion on a way around this?

File: lib/modules/manager/flux/extract.ts
Comment (viceice) [inline lib/modules/manager/flux/extract.ts:212]:
```suggestion
        const dep: PackageDependency = {
```

**Review by viceice — COMMENTED** [review]
you can probably also remove that from deps and return type.

**Comment by onedr0p** [comment]
@viceice should be all set again

**Comment by onedr0p** [comment]
@viceice any chance on another review? I am keen to get started on some other work for this manager.

**Review by viceice — APPROVED** [review]

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the docs.

**Pull request merged**
